### PR TITLE
Chore refactor columns to redux reselect

### PR DESCRIPTION
**Pull request by kostadriano**
This PR is a refactor in Redux structure of board V2. This was necessary for two reasons:
- Stories state was duplicated out and inside the columns, so any action in the determined story need to be dispatch in all columns.
- All stories were putting one by one in columns, so the sort algorithm was running several times.

So instead of putting stories inside a column state, it was all filtered using Redux [reselect](https://github.com/reduxjs/reselect#selectorsindexjs). Here we can see the state tree difference.

before:
image

after:
image

This refactoring will help the next features.

### CODE REVIEW

File: app/assets/javascripts/selectors/done.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import moment from "moment";
+
+export const mountPastIterations = (pastIterations) => {
+  const sprints = pastIterations.map((sprint, index) => ({
+    ...sprint,
+    number: index + 1,
+    startDate: moment(sprint.startDate).format("ddd MMM Do Y"),
```
Comment (lucasdeassis) [inline app/assets/javascripts/selectors/done.js]: depending on the locale assigned (pt for exemple), the formal output may be something like `Qui dez 6º 2018​​​​​`, in which looks a bit weird. What do you think?

Also, IMO it is not good to add/deal with `moment` or any text formatting in redux selectors. I feel that it is something more react component related and might limit past iterations formatting later. I believe that `startDate` and `endDate` should be in ISO here

File: app/assets/javascripts/selectors/columns.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+const getStories = (state) => state.stories;
+const getColumn = (state) => state.column;
+const getProject = (state) => state.project;
+const getPastIterations = (state) => state.pastIterations;
+
+export const getColumns = createSelector(
+  [getColumn, getStories, getProject, getPastIterations],
+  (column, stories, project, pastIterations) => {
+    switch (column) {
+      case actionTypes.COLUMN_CHILLY_BIN:
```
Comment (talyssonoc) [inline app/assets/javascripts/selectors/columns.js]: Are these actual action types? Or just constants with the ids of the columns? I think we could move these constants out of `actionTypes.js`, what do you think?

File: spec/javascripts/reducers/stories_spec.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  }
+];
+
+describe('Stories reducer', () => {
+  describe("Toggle story", () => {
+    describe("When story is collapsed", () => {
+      it("expand story", () => {
+        const initialState = storiesArray;
+        const story = storiesArray[0];
+        story.collapsed = true;
```
Comment (talyssonoc) [inline spec/javascripts/reducers/stories_spec.js]: This is gonna mutate the story from the props, making a test share state with the others. You could do something like this:

```js
describe('Stories reducer', () => {
  let storiesArray;

  beforeEach(() => {
    storiesArray = [
      {
        id: 1,
        collapsed: true
      },
      {
        id: 2,
        collapsed: true
      },
      {
        id: 3,
        collapsed: true
      }
    ];
  });
  describe("Toggle story", () => {
    describe("When story is collapsed", () => {
      it("expand story", () => {
```
And then on every test it's gonna be a new array so no risk on mutating a story from the array 😄

File: spec/javascripts/selectors/backlog_spec.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+        const stories = [newStory, oldStory];
+        const orderedStories = orderByState(stories);
+        
+        expect(orderedStories[0]).toEqual(oldStory);
+        expect(orderedStories[1]).toEqual(newStory);
+      });
+    });
+
+    describe('Story state is delivered', () => {
```
Comment (talyssonoc) [inline spec/javascripts/selectors/backlog_spec.js]: `Story state is delivered` -> `Are stories are delivered`, what do you think?

Comment (talyssonoc) [reply spec/javascripts/selectors/backlog_spec.js]: The same can be done with the other tests.

File: spec/javascripts/selectors/columns_spec.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    const storyAcceptedCurrentSprint = storyFactory({ state: 'accepted', acceptedAt: currentSprintDate });
+
+    const project = {
+      iterationLength: 1,
+      iterationStartDay: 1,
+      startDate: currentSprintDate.format()
+    };
+
+    it('return story accepted in current sprint', () => {
+      storiesArray.push(storyAcceptedCurrentSprint);
```
Comment (talyssonoc) [inline spec/javascripts/selectors/columns_spec.js]: You're mutating shared state here as well, we could use the same approach for the other comment. Using `let`  + `beforeEach`

File: app/assets/javascripts/selectors/done.js (hunk 1 above)
Comment (kostadriano) [reply app/assets/javascripts/selectors/done.js]: Thank for the review, Lucas! Yep, great idea, I gonna do this in another PR to change the dates in considering of user locale removing this logic to React component.

Comment (kostadriano) [reply app/assets/javascripts/selectors/done.js]: Ref: [23314](https://central.cm42.io/projects/central-board-v2#story-23314)

File: spec/javascripts/selectors/columns_spec.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+describe('Columns Selector', () => {
+  const unscheduledStory = storyFactory({ state: 'unscheduled' });
+  const startedStory = storyFactory({ state: 'started' });
+  const deliveredStory = storyFactory({ state: 'delivered' });
+
+  const currentSprintDate = moment();
+  const previousSprintDate = moment().subtract(1, 'weeks');
+
+  let storiesArray
```
Comment (talyssonoc) [inline spec/javascripts/selectors/columns_spec.js]: Missing a `;` here 😛

**Pull request merged**